src/rest/connection.cpp

```diff
@@ -0,0 +1,364 @@
+            {
+                mResource->Handle(mRequest, mResponse);
+
+                if (mResponse.NeedCallback())
+                {
+                    // Set its state as Waitforcallback and refresh its timer
+                    mState     = OTBR_REST_CONNECTION_CALLBACKWAIT;
+                    mStartTime = steady_clock::now();
+                }
+                else
+                {
+                    // Normal Write back process.
+                    Write();
+                }
+            }
+            else
+            {
+                // 1. Catch the error
+                // 2. Client side has close this connection(received == 0).
+                VerifyOrExit(received > 0 || (received == -1 && (err == EAGAIN || err == EWOULDBLOCK)),
+                             error = OTBR_ERROR_REST);
+            }
+        }
+    }
+    else
+    {
+        Disconnect();
+    }
+
+exit:
+    if (error == OTBR_ERROR_REST)
+    {
+        Disconnect();
+    }
+    if (mState <= OTBR_REST_CONNECTION_READWAIT)
+    {
+        mState = OTBR_REST_CONNECTION_READWAIT;
+    }
+    return error;
+}
+
+otbrError Connection::ProcessWaitCallback()
+{
+    otbrError error    = OTBR_ERROR_NONE;
+    auto      duration = duration_cast<microseconds>(steady_clock::now() - mStartTime).count();
+
+    if (duration >= kCallbackTimeout)
+    {
+        // Handle Callback, then Write back response
+        mResource->HandleCallback(mRequest, mResponse);
+        Write();
+    }
+
+    return error;
+}
+
+otbrError Connection::ProcessWaitWrite(fd_set &aWriteFdSet)
+{
+    otbrError error    = OTBR_ERROR_NONE;
+    auto      duration = duration_cast<microseconds>(steady_clock::now() - mStartTime).count();
+
+    if (duration <= kWriteTimeout && FD_ISSET(mFd, &aWriteFdSet))
+    {
+        // WriteFdSet is set, then try to write again
+        Write();
+    }
+    else
+    { // Pass write timeout, just close the error
+        Disconnect();
+    }
+    return error;
+}
+
+otbrError Connection::Process(fd_set &aReadFdSet, fd_set &aWriteFdSet)
+{
+    otbrError error = OTBR_ERROR_NONE;
+
+    switch (mState)
+    {
+    // Initial state, directly read for the first time.
+    case OTBR_REST_CONNECTION_INIT:
+    {
+        error = ProcessWaitRead(aReadFdSet);
+        break;
+    }
+    case OTBR_REST_CONNECTION_READWAIT:
+    {
+        error = ProcessWaitRead(aReadFdSet);
+        break;
+    }
+    case OTBR_REST_CONNECTION_CALLBACKWAIT:
+    {
+        //  Wait for Callback process.
+        error = ProcessWaitCallback();
+        break;
+    }
+    case OTBR_REST_CONNECTION_WRITEWAIT:
+    {
+        error = ProcessWaitWrite(aWriteFdSet);
+        break;
+    }
+    default:
+        break;
+    }
+
+    return error;
+}
+
+otbrError Connection::Write()
+{
+    otbrError   error = OTBR_ERROR_NONE;
+    std::string errorCode;
+    int32_t     sendPointer;
+    int         err;
+
+    if (mState < OTBR_REST_CONNECTION_WRITEWAIT)
```

Comment:
You should not use comparison operator on `enum` values!

Reply:
Write a short function for this, with a good name! Otherwise, nobody knows the meaning of `mState < OTBR_REST_CONNECTION_WRITEWAIT`.